jib-build-plan/src/main/java/com/google/cloud/tools/jib/api/buildplan/ContainerBuildPlan.java

```diff
@@ -0,0 +1,495 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.api.buildplan;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+/** Describes a plan to build a container. */
+@Immutable
+public class ContainerBuildPlan {
+
+  /** Builder for {@link ContainerBuildPlan}. */
+  public static class Builder {
+
+    private String baseImage = "scratch";
+    private String architectureHint = "amd64";
+    private String osHint = "linux";
+    private Instant creationTime = Instant.EPOCH;
+    private ImageFormat format = ImageFormat.Docker;
+
+    // image execution parameters
+    private Map<String, String> environment = new HashMap<>();
+    private Map<String, String> labels = new HashMap<>();
+    private Set<AbsoluteUnixPath> volumes = new HashSet<>();
+    private Set<Port> exposedPorts = new HashSet<>();
+    @Nullable private String user;
+    @Nullable private AbsoluteUnixPath workingDirectory;
+    @Nullable private List<String> entrypoint;
+    @Nullable private List<String> cmd;
+
+    private List<LayerObject> layers = new ArrayList<>();
```

Comment:
Not sure I like the name `LayerObject`, but that's a different discussion.

Reply:
I don't like it either. We already have a non-public `Layer` interface in jib-core, so I just wanted to avoid name clash.

Reply:
Okay, well this shouldn't prevent this PR from going in.